internal/internal.go

```diff
@@ -422,27 +422,41 @@ func (p *envoyExtAuthzGrpcServer) check(ctx context.Context, req interface{}) (*
 
 	input, err = envoyauth.RequestToInput(req, logger, p.cfg.protoSet, p.cfg.SkipRequestBodyParse)
 	if err != nil {
-		internalErr = internalError(RequestParseErr, err)
-		return nil, stop, &internalErr
+		internalErr = newInternalError(RequestParseErr, err)
+		return &ext_authz_v3.CheckResponse{
+			Status: &rpc_status.Status{
+				Code:    int32(code.Code_INVALID_ARGUMENT),
+				Message: internalErr.Error(),
+			},
+			HttpResponse: &ext_authz_v3.CheckResponse_DeniedResponse{
+				DeniedResponse: &ext_authz_v3.DeniedHttpResponse{
+					Status: &ext_type_v3.HttpStatus{
+						Code: ext_type_v3.StatusCode(ext_type_v3.StatusCode_BadRequest),
+					},
+					Body: internalErr.Error(),
+				},
+			},
+			DynamicMetadata: nil,
+		}, stop, internalErr
```

Comment:
I think we need to verify here if Envoy returns `403` by default or `400`. It's possible if you return an error along with the response, `403` will be the default. Would be good to verify this.

Reply:
> Looks like setting the DeniedHttpResponse.Status field allows the authz service to send an HTTP response to the downstream client.

Sure but what happens when the authz service returns both an error and a response. Have you verified the status from the response is picked up?

Reply:
> Code pointers:
[v3/external_auth.pb.go#L280-L283](https://github.com/envoyproxy/go-control-plane/blob/main/envoy/service/auth/v3/external_auth.pb.go#L280-L283)
[v3/external_auth.pb.go#L83-L85](https://github.com/envoyproxy/go-control-plane/blob/main/envoy/service/auth/v3/external_auth.pb.go#L83-L85)

Looks like the response status will be picked up but good to verify in any case.

Reply:
On a second glance looks like my `CheckResponse.Status` must be `403` but `DeniedHttpResponse.Status` can be `401`. That case Envoy treats it as a denied request but sends `401` HTTP status code downstream. WDYT @ashutosh-narkar ?

Reply:
I think `CheckResponse.Status` can be any non-200 and it will considered as deny. `DeniedHttpResponse.Status` overrides it and is sent downstream. So if you just set `CheckResponse.Status` to say `400` will Envoy still send `403` to the client? If no, then you can probably avoid `DeniedHttpResponse`.

Reply:
I modified the PR, the `CheckResponse.Status` is `200 OK` but downstream response in `403 Forbidden`. Do checkout the links above and let me know WYT.

Reply:
@ashutosh-narkar we can't set `CheckResponse.Status` as non-200, this is because for all non-OK responses `status_on_error` sends default `403`. Check [this](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto#L96) out.

Reply:
> @ashutosh-narkar we can't set CheckResponse.Status as non-200, this is because for all non-OK responses status_on_error sends default 403.

But you can override it in the denied response. I think the way you had it originally, ie `403` for `CheckResponse.Status` and `400` for `DeniedHttpResponse.Status` seemed correct to me.

Reply:
@ashutosh-narkar I went through envoy control plane docs. I think the right thing to do would be as suggested by you, return a `403` CheckResponse Status and a `401` HTTP downstream status. The error should be nil though. Updated the PR accordingly. Please review. Thanks!